util/dao/src/lib.rs

```diff
@@ -1,50 +1,310 @@
+        let added_occupied_capacities =
+            rtxs.iter().try_fold(Capacity::zero(), |capacities, rtx| {
+                rtx.transaction
+                    .outputs()
+                    .iter()
+                    .try_fold(Capacity::zero(), |tx_capacities, output| {
+                        output
+                            .occupied_capacity()
+                            .and_then(|c| tx_capacities.safe_add(c))
+                    })
+                    .and_then(|c| capacities.safe_add(c))
+            })?;
+
+        let (parent_ar, parent_c, parent_u) = extract_dao_data(parent.dao())?;
+
+        let parent_g2 = calculate_g2(
+            parent.number(),
+            &parent_epoch,
+            self.consensus.secondary_epoch_reward(),
+        )?;
+        let parent_g = if parent.number() > self.consensus.reserve_number() {
+            // Note the primary issuance of a block is not the reward of the
+            // parent block, but the target block(see RewardCalculator)
+            let target_number = self
+                .consensus
+                .finalize_target(parent.number())
+                .ok_or(Error::InvalidHeader)?;
+            let target = self
+                .store
+                .get_ancestor(parent.hash(), target_number)
+                .ok_or(Error::InvalidHeader)?;
+            self.primary_block_reward(&target)
+                .and_then(|c| c.safe_add(parent_g2).map_err(Into::into))?
+        } else {
+            self.consensus
+                .genesis_epoch_ext()
+                .block_reward(parent.number())?
+        };
+
+        let current_c = parent_c.safe_add(parent_g)?;
+        let current_ar = u128::from(parent_ar)
+            * u128::from((parent_c.safe_add(parent_g2)?).as_u64())
+            / (max(u128::from(parent_c.as_u64()), 1));
+        let current_u = parent_u
+            .safe_add(added_occupied_capacities)
+            .and_then(|u| u.safe_sub(freed_occupied_capacities))?;
+
+        Ok(pack_dao_data(current_ar as u64, current_c, current_u))
+    }
 
-    let withdraw_capacity =
-        Capacity::shannons(withdraw_counted_capacity as u64).safe_add(occupied_capacity)?;
-    Ok(withdraw_capacity)
+    pub fn maximum_withdraw(
+        &self,
+        out_point: &OutPoint,
+        withdraw_header_hash: &H256,
+    ) -> Result<Capacity, FailureError> {
+        let cell_out_point = out_point.cell.as_ref().ok_or(Error::InvalidOutPoint)?;
+        let (tx, block_hash) = self
+            .store
+            .get_transaction(&cell_out_point.tx_hash)
+            .ok_or(Error::InvalidOutPoint)?;
+        let output = tx
+            .outputs()
+            .get(cell_out_point.index as usize)
+            .ok_or(Error::InvalidOutPoint)?;
+        self.calculate_maximum_withdraw(&output, &block_hash, withdraw_header_hash)
+    }
+
+    pub fn transaction_fee(&self, rtx: &ResolvedTransaction) -> Result<Capacity, FailureError> {
```

Comment:
This indeed will also verify DAO withdraw. It is a looser version of the dao.c script in the system cell.